pydantic/main.py

```diff
@@ -75,6 +77,42 @@ def __pydantic_post_init__(self_: Any, **kwargs: Any) -> None:
             validator_functions = _validation_functions.ValidationFunctions(bases)
             namespace['__pydantic_validator_functions__'] = validator_functions
 
+           # Errors for misuse of discriminated unions (check test_discriminated_unions)
+            for name, value in namespace['__annotations__'].items():
+                #Ensure value is intended as a discriminated union
+                if isinstance(namespace.get(name), FieldInfo) and namespace[name].discriminator is not None:
+                    # If the type expected is not a union or is a single value union, raise TypeError
+                    if get_origin(value) is not Union:
+                        raise TypeError('`discriminator` can only be used with `Union` type with more than one variant')
+                    
+                    for arg in get_args(value):
+                        # print(get_origin(arg))
+                        if get_origin(arg) is Annotated:
+                            arg = get_args(arg)[0]
+                            # arg = get_args(arg)
+                            for nested_type in get_args(arg):
+                                if not issubclass(nested_type, BaseModel) and not is_dataclass(nested_type):
+                                    raise TypeError(f"Type '{nested_type.__name__}' is not a valid `BaseModel` or `dataclass`")
+                        elif not issubclass(arg, BaseModel) and not is_dataclass(arg):
+                            raise TypeError(f"Type '{arg.__name__}' is not a valid `BaseModel` or `dataclass`")
+                    
+                    # Ensure discriminated value is present in all union options, if not raise error
```

Comment:
this is not long required, the key doesn't need to exist in all versions of the union.